AWSutils.sh

```diff
@@ -0,0 +1,509 @@
+function check_amzn_rpms()
+{
+   mapfile -t AMZNRPMS < <( stat -c '%n' "${UTILSDIR}"/*.el7.*.rpm )
+   if [[ ${#AMZNRPMS[@]} -eq 0 ]]
+   then
+      (
+      echo "AMZN.Linux packages not found in ${UTILSDIR}"
+      echo "Please download missing RPMs before proceeding."
+      echo "Note: GetAmznLx.sh may be used to do this for you."
+      echo "Aborting..."
+      ) > /dev/stderr
+      exit 1
+   fi
+}
+
+# Enable the RHEL "optional" repo where appropriate
+function enable_rhel_optional_repo()
+{
+    OPTIONREPO=$(yum repolist all | grep rhel-server-optional || true)
+    if [[ -n ${OPTIONREPO} ]]
+    then
+        chroot "${CHROOTMNT}" yum-config-manager --enable "${OPTIONREPO/\/*/}"
+    fi
+}
+
+# Force systemd services to be enabled in resultant AMI
+function enable_services()
+{
+    for SVC in "${SYSTEMDSVCS[@]}"
+    do
+        printf "Attempting to enable %s in %s... " "${SVC}" "${CHROOTMNT}"
+        chroot "${CHROOTMNT}" /usr/bin/systemctl enable "${SVC}" || err_exit "FAILED"
+        echo "SUCCESS"
+    done
+}
+
+# Get list of rpm filenames
+function get_awstools_filenames()
+{
+    if [[ -z ${AWSTOOLSRPM:-} ]]
+    then
+        ls "${UTILSDIR}"/*.el7.*.rpm
+    else
+        rpmfiles=""
+        for rpmfile in ${AWSTOOLSRPM}
+        do
+            rpmfiles="${rpmfiles} "$(ls "${UTILSDIR}/${rpmfile}"*.el7.*.rpm)
+        done
+        echo "${rpmfiles}"
+    fi
+}
+
+# Install aws utils rpms
+function install_aws_utils()
+{
+    # Depending on RPMs dependencies, this may fail if a repo is
+    # missing (e.g. EPEL). Will also fail if no RPMs are present
+    # in the search directory.
+    # use yum info .. to check and show details for installed RPMs.
+    rpmfiles=$(get_awstools_filenames)
+    if [[ -z ${rpmfiles} ]]
+    then
+        echo "No Installation of additional Amazon RPMs"
+    else
+        # shellcheck disable=2086
+        yum --installroot="${CHROOTMNT}" install -e 0 -y ${rpmfiles} || exit $?
+        enable_services
+    fi
+}
+
+# Install AWS CLI version 1.x
+function InstallCLIv1 {
+   local INSTALLDIR
+   local BINDIR
+
+   INSTALLDIR="/usr/local/aws-cli/v1"
+   BINDIR="/usr/local/bin"
+
+   if [[ ${CLIV1SOURCE} == "UNDEF" ]]
+   then
+      err_exit "AWS CLI v1 not requested for install. Skipping..." NONE
+   elif [[ ${CLIV1SOURCE} == http[s]://*zip ]]
+   then
+      # Make sure Python3 is present
+      EnsurePy3
+
+      err_exit "Fetching ${CLIV1SOURCE}..." NONE
+      curl -sL "${CLIV1SOURCE}" -o "${CHROOTMNT}/tmp/awscli-bundle.zip" || \
+        err_exit "Failed fetching ${CLIV1SOURCE}"
+
+      err_exit "Dearchiving awscli-bundle.zip..." NONE
+      (
+         cd "${CHROOTMNT}/tmp"
+         unzip -q awscli-bundle.zip
+      ) || \
+        err_exit "Failed dearchiving awscli-bundle.zip"
+
+      err_exit "Installing AWS CLIv1..." NONE
+      chroot "${CHROOTMNT}" /bin/bash -c "/tmp/awscli-bundle/install -i '${INSTALLDIR}' -b '${BINDIR}/aws'" || \
```

Comment:
Doesn't the folder parameter for `-I` here also need `/aws` at the end in v1?
```suggestion
      chroot "${CHROOTMNT}" /bin/bash -c "/tmp/awscli-bundle/install -i '${INSTALLDIR}/aws' -b '${BINDIR}/aws'" || \
```

Reply:
No, the logic is accounting for differences in the install routine for v1 vs v2... INSTALLDIR is `/usr/local/aws-cli/v1`, which tells the v1 installer to put the `aws` binary at `/usr/local/aws-cli/v1/bin/aws`. The -b option is also slightly different, as it is the full path _filename_ you want for a linked binary.

In v2, -i controls a top-level directory and the installer creates versioned directories within that location, at `<INSTALLDIR>/v2/<version>`. And the binary is placed at `v2/<version>/bin/aws`. It also creates a <latest> symlink to the <version> directory, which is what we use to create the `aws2` symlink. The -b option is a _directory_ where the installer will place the `aws` linked binary, i.e. you cannot control the filename with the v2 installer.

Reply:
Got it.  Thanks for the clarification.